src/body/tee_body.rs

```diff
@@ -0,0 +1,163 @@
+use bytes::{Buf, Bytes};
+use http_body::{Body, Frame};
+use std::collections::VecDeque;
+use std::fmt::{self, Debug, Formatter};
+use std::io::{self, ErrorKind};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::AsyncWrite;
+
+use super::BoxBody;
+use crate::error::DynError;
+
+/// A boxed body that writes each data frame into a dyn
+/// [`AsyncWrite`](AsyncWrite).
+///
+pub struct TeeBody {
+    state: IoState,
+    io: Pin<Box<dyn AsyncWrite + Send + Sync>>,
+    body: BoxBody,
+    backlog: VecDeque<Bytes>,
+}
+
+enum IoState {
+    Closed,
+    Shutdown,
+    Writeable,
+}
+
+fn broken_pipe() -> DynError {
+    Box::new(io::Error::from(ErrorKind::BrokenPipe))
+}
+
+impl TeeBody {
+    pub fn new(body: BoxBody, io: impl AsyncWrite + Send + Sync + 'static) -> Self {
+        Self {
+            io: Box::pin(io),
+            body,
+            state: IoState::Writeable,
+            backlog: VecDeque::with_capacity(2),
+        }
+    }
+}
+
+impl Body for TeeBody {
+    type Data = Bytes;
+    type Error = DynError;
+
+    fn poll_frame(
+        self: Pin<&mut Self>,
+        context: &mut Context<'_>,
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+        let this = self.get_mut();
+
+        let io = &mut this.io;
+        let state = &mut this.state;
+        let backlog = &mut this.backlog;
+        let mut next = None;
+
+        loop {
+            match state {
+                IoState::Writeable => {}
```

Comment:
I'm having a hard time dealing with this empty match arm here. Perhaps basing the state on the shutdown rather than the write-ability of `io` would make my intentions more clear?

I'm going to brainstorm a bit, write some tests, and then mark this as ready for review and merge asap to get another release out along with a couple refactors to the body module in a follow up PR.